### DISCUSSION TITLE
Graphs plots in 3-D

### DISCUSSION THREAD

**Discussion by Florian2Richter** [networkx/networkx#6803] [Ideas]
Hello everyone,

As a newcomer to both open-source projects and the world of graphs, I must admit that I might lack the experience some of you have here. Due to my limited experience, I approach my proposal with humility, realizing that my idea may need refinement or may not be entirely correct.

Having said this, I wonder whether it makes sense to expand the functionality of nx.draw_networkx() to three dimensions. After looking at the code, it seems that this expansion might not require too much effort. While I understand that NetworkX is not meant to be a fully-featured graph visualization tool, I'd like to present my points on why it might make sense to consider this extension:

-  **Existing Layout Functions Support:** I noticed that most of the layout functions (e.g., random layout, spring layout) seem to support a third dimension (dim=3) already. This indicates that NetworkX might already have some groundwork in place to explore the third dimension.

-     **Algorithmic Differences:** Graph layouts in three dimensions can produce different results compared to their two-dimensional counterparts. Visualizing graphs in 3D may reveal unique insights and patterns that might be hidden in traditional 2D representations.

-     **Implementation Feasibility:** The effort required for implementing 3D visualization in matplotlib, the underlying plotting library used by NetworkX, might not be overly demanding. The integration of additional dimensions could enhance the overall visualization experience.

I'd like to share some examples to illustrate my point further. Here are links to two images showcasing a graph representing football data in both 2D and 3D:
![2_d_football](https://github.com/networkx/networkx/assets/95981719/7d5d88a9-4e19-408b-ba81-6afb179eb571)
![3_d_football](https://github.com/networkx/networkx/assets/95981719/61fae166-7216-431a-ba5a-f277de5b6ba8)
As you can observe, the 3D representation appears to provide a clearer visualization of the yellow cluster compared to the 2D version.

I genuinely value your feedback on whether extending the plotting abilities to include a third dimension would be worthwhile. I am eager to learn from your insights and expertise in this field. If this idea aligns with the goals of NetworkX and has potential merit, I would be excited to contribute and work towards its implementation.

Thank you all for considering my proposal, and I'm looking forward to your valuable input.

Best regards,
Florian

**Comment by rossbar** [comment]
> whether extending the plotting abilities to include a third dimension would be worthwhile

This is already possible: see e.g. [this section in the gallery](https://networkx.org/documentation/latest/auto_examples/index.html#d-drawing). I suppose the main thing is that NetworkX doesn't provide any specific API for plotting in 3D.

I suppose the question is: what specifically is the proposal? Is it to add something like `if pos.ndim == 3: ax = Axes3D` to the existing `nx_pylab` drawing functions? This *might* work, though it will require some exploration to see how much of the 2D functionality directly maps to 3D. If it's to add a different plotting backend (i.e. something other than matplotlib), my sense is that may make more sense to work on in an external project - especially since (IME) full-feature 3D plotting often requires a larger suite of dependencies.

It wouldn't hurt to expand the example gallery though!

**Reply by MridulS** [reply]
Plus there is still a limitation of FancyArrow class with 3d vectors https://github.com/networkx/networkx/pull/4489#issuecomment-1637168376 so having a NetworkX wide drawing API for 3d plots could be challenging (but definitely worthwhile!)

**Reply by Florian2Richter** [reply]
Hi All, 
thanks for hinting to the 3-D examples, and yes, the idea would be just add a pos.ndim == 3 variable to the existing plotting functions (e.g. nx.draw_networkx()). Everything beyond that (new plotting backend, etc.) seems to be out of the scope for Networkx, as there are other existing solutions. 
As you mentioned the examples section, I found time also to generate the 3_d Version of the Facebook example:
![facebook_2d_compressed](https://github.com/networkx/networkx/assets/95981719/aaa35631-8a67-4d35-9e39-945f5c7dc8a0)
![facebook_3d_compressed](https://github.com/networkx/networkx/assets/95981719/b50b37f0-42b1-4fec-b504-1f92474fd24e)

So, my proposal would be to open a ticket with the API extension to 3-D for the existing functions and another ticket for generating/updating the examples section. Goes without saying, that I will take up do the best I can to contribute to those...

What do you think about that?

**Reply by rossbar** [reply]
> What do you think about that?

FWIW there have been attempts in the past - see e.g. #4489 - that have been blocked by various limitations with how much of the nx drawing API maps to 3D matplotlib functionality. It's worth revisiting those conversations to see if they can inform another attempt!

**Comment by Florian2Richter** [comment]
Since I am struggling a bit to find time for the general implementation, I would first add those examples to the gallery... Please let me know, if you require an issue to be opened for that. Otherwise, I start direct along the guidelines of the FAQ...

**Reply by dschult** [reply]
You don't need to create an "Issue", just open a PR.  This discussion and comments on that PR will provide sufficient info to review/merge it. :)
